### PR TITLE
Scrabble

### PR DESCRIPTION
**Pull request by MonalisaC**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

|  Feature 	|   Feedback	|
|---	|---	|
|   How do you feel you and your partner did in sharing responsibilities?	|  well 	|
|   Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? |   So that we can call this method from other classes and it remains independent of classes we are calling it from. |
|   Describe an example of a test you wrote for a _nominal case_. |Should remove the tiles from the default set in tilebag class's draw tiles method
 |
|   Describe an example of a test you wrote for an _edge case_. |  Raises an error when the word parameter is invalid data type eg. ("sammy3") in Scoring class's score method. 	|
|   What was your final test coverage percentage?	| 100% except except player_spec  	|
|   Is there a specific piece of code you'd like feedback on?	| We need some help coming up with edge case test for draw tiles test method  in player class  .	|

### CODE REVIEW

File: lib/player.rb [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    end
+
+    def highest_score_word
+      # Calls the class method from Scrabble::Scoring
+      return Scoring.highest_score_from(@plays)
+    end
+
+    def highest_word_score
+      # Calls the class method from Scrabble::Scoring within the loop
+      scores = @plays.map{ |word| Scoring.score(word) }
```
Comment (CheezItMan) [inline lib/player.rb:54]: You could use `highest_score_from` from the `Scoring` class.

File: specs/player_spec.rb [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    end
+
+    it "Validates the argument" do
+      proc{
+        @player.play(12345)
+      }.must_raise ArgumentError
+      proc{
+        @player.play("qwe12")
+      }.must_raise ArgumentError
+    end
```
Comment (CheezItMan) [inline specs/player_spec.rb:31]: you should probably also include tests using the empty string. ("")

File: specs/tilebag_spec.rb [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+      }.must_raise ArgumentError
+    end
+
+    it "Returns the number of tiles remaining in the bag" do
+      @tilebag.draw_tiles(7)
+      @tilebag.tiles_remaining.must_equal 91
+      @tilebag.tiles.length.must_equal 91
+      @tilebag.draw_tiles(91)
+      @tilebag.tiles_remaining.must_equal 0
+      @tilebag.tiles.length.must_equal 0
```
Comment (CheezItMan) [inline specs/tilebag_spec.rb:71]: You should test what happens when you draw too many tiles!

**Comment by CheezItMan** [comment]
# Scrabble
## What We're Looking For

|  Feature 	|   Feedback |
|---	|---	|
|  **General**	|   	|
|   Answered comprehension questions 	|   Maybe put a little bit more than "well" in your answer...  `score` is a method that doesn't apply to a particular player or tile etc.  It simply gets called and calculates an answer.  	|
|   Both Teammates contributed to the codebase	|   Good job sharing the coding and committing.  	|
|   Regular Commits with meaningful commit messages 	|   Good regular committing	|
|  **Scoring class**	|   	|
|  `score` method	|   Check, good use of enumerables	|
|   Uses appropriate data structure to store the letter score	|   Good use of a hash	|
|   Appropriately handles edge cases	|   Checks invalid words, and 7-letter word scores	|
|   Tests `score` edge cases (one-letter word, 7 letter word)	|   Checks 7-letter word	|
|  `highest_scoring_word` method	|   Check	|
|   Appropriately handles edge cases	|   Check	|
|   Tests `highest_scoring_word` edge cases (ties, no words, no array)	|   Check	|
|  **Player class**	|   	|
|  Uses an array to keep track of words played	|   Check	|
|  Uses existing code to create `highest_scoring_word` and `highest_word_score`	|   Check	|
|  Returns `true` or `false` from `won?` method |  Good use of ternary operator |
|  Tests edge cases on the `play(word)` method for words that cannot be scored |   Missing tests for words that are the emptyString  |
|  **TileBag class**	|   	|
|  Uses a data structure to keep the set of default tiles in the bag	|   Good use of an array	|
|  `draw_tiles` method uses the parameter to draw a set of random tiles |   Good idea with `shuffle!` & `shift`	|
| `tiles_remaining` method returns a count	|    Check	|
|  Tests `draw_tiles` edge cases for different parameter inputs	|   Missing tests on when more tiles are being drawn than exist	|


## Summary

Very nice work.  You've covered the requirements.  Your testing could use some improvements but it was pretty good.  It takes experience to get good at knowing what to test.  Very good use of enumerables.

**Review by CheezItMan — COMMENTED** [review]
Very nice work.  You've covered the requirements.  Your testing could use some improvements but it was pretty good.  It takes experience to get good at knowing what to test.  Very good use of enumerables.
